File: common/src/main/java/net/impleri/playerskills/integration/ftbquests/NumericSkillReward.java
package net.impleri.playerskills.integration.ftbquests;

import dev.ftb.mods.ftblibrary.config.ConfigGroup;
import dev.ftb.mods.ftbquests.quest.Quest;
import dev.ftb.mods.ftbquests.quest.reward.RewardType;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.impleri.playerskills.api.SkillType;
import net.impleri.playerskills.registry.RegistryItemNotFound;
import net.impleri.playerskills.server.ServerApi;
import net.impleri.playerskills.server.api.Skill;
import net.impleri.playerskills.variant.numeric.NumericSkillType;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.server.level.ServerPlayer;

public class NumericSkillReward extends BasicSkillReward {
    protected static double NO_VALUE = -1.0;
    public double min = NO_VALUE;
    public double max = NO_VALUE;

    public NumericSkillReward(Quest quest) {
        super(quest);
    }

    @Override
    protected ResourceLocation getSkillType() {
        return NumericSkillType.name;
    }

    @Override
    public RewardType getType() {
        return SkillRewardTypes.NUMERIC_SKILL;
    }

    @Override
    public void writeData(CompoundTag nbt) {
        super.writeData(nbt);

        nbt.putDouble("min", min);
        nbt.putDouble("max", max);
    }

    @Override
    public void readData(CompoundTag nbt) {
        super.readData(nbt);

        min = nbt.getDouble("min");
        max = nbt.getDouble("max");
    }

    @Override
    public void writeNetData(FriendlyByteBuf buffer) {
        super.writeNetData(buffer);

        buffer.writeDouble(min);
        buffer.writeDouble(max);
    }

    @Override
    public void readNetData(FriendlyByteBuf buffer) {
        super.readNetData(buffer);

        min = buffer.readDouble();
        max = buffer.readDouble();
    }

    @Override
    @Environment(EnvType.CLIENT)
    public void getConfig(ConfigGroup config) {
        super.getConfig(config);

        config.addDouble("min", min, v -> min = v, NO_VALUE, NO_VALUE, Double.MAX_VALUE)
                .setNameKey("playerskills.quests.ui.min");

        config.addDouble("max", max, v -> max = v, NO_VALUE, NO_VALUE, Double.MAX_VALUE)
                .setNameKey("playerskills.quests.ui.max");
    }

    @Override
    public void claim(ServerPlayer player, boolean notify) {
        net.impleri.playerskills.api.Skill<Double> actualSkill;
        SkillType<Double> skillType;
        Double nextVal;

        try {
            actualSkill = Skill.find(skill);
            skillType = SkillType.find(getSkillType());
        } catch (RegistryItemNotFound e) {
            throw new RuntimeException(e);
        }

        nextVal = skillType.getNextValue(actualSkill, min < 0 ? null : min, max < 0 ? null : max);
        if (downgrade) {
            nextVal = skillType.getPrevValue(actualSkill, min < 0 ? null : min, max < 0 ? null : max);
        }

        if (nextVal != null && ServerApi.set(player, skill, nextVal)) {
            maybeNotify(player, notify, nextVal.toString());
        }
    }
}

File: common/src/main/kotlin/net/impleri/playerskills/integrations/ftbquests/NumericSkillReward.kt
package net.impleri.playerskills.integrations.ftbquests

import dev.ftb.mods.ftblibrary.config.ConfigGroup
import dev.ftb.mods.ftbquests.quest.Quest
import dev.ftb.mods.ftbquests.quest.reward.RewardType
import net.fabricmc.api.EnvType
import net.fabricmc.api.Environment
import net.impleri.playerskills.api.Player
import net.impleri.playerskills.api.SkillType
import net.impleri.playerskills.skills.numeric.NumericSkillType
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.FriendlyByteBuf
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerPlayer

open class NumericSkillReward(quest: Quest?) : BasicSkillReward(quest) {
  var min = NO_VALUE
  var max = NO_VALUE
  override val skillType: ResourceLocation
    get() = NumericSkillType.NAME

  override fun getType(): RewardType {
    return SkillRewardTypes.NUMERIC_SKILL
  }

  override fun writeData(nbt: CompoundTag) {
    super.writeData(nbt)
    nbt.putDouble("min", min)
    nbt.putDouble("max", max)
  }

  override fun readData(nbt: CompoundTag) {
    super.readData(nbt)
    min = nbt.getDouble("min")
    max = nbt.getDouble("max")
  }

  override fun writeNetData(buffer: FriendlyByteBuf) {
    super.writeNetData(buffer)
    buffer.writeDouble(min)
    buffer.writeDouble(max)
  }

  override fun readNetData(buffer: FriendlyByteBuf) {
    super.readNetData(buffer)
    min = buffer.readDouble()
    max = buffer.readDouble()
  }

  @Environment(EnvType.CLIENT)
  override fun getConfig(config: ConfigGroup) {
    super.getConfig(config)
    config.addDouble("min", min, { min = it }, NO_VALUE, NO_VALUE, Double.MAX_VALUE)
      .setNameKey("playerskills.quests.ui.min")
    config.addDouble("max", max, { max = it }, NO_VALUE, NO_VALUE, Double.MAX_VALUE)
      .setNameKey("playerskills.quests.ui.max")
  }

  override fun claim(player: ServerPlayer, notify: Boolean) {
    val actualSkill = skill?.let { Player.get<Double>(player, it) } ?: throw RuntimeException()
    val skillType = actualSkill.let { SkillType.find<Double>(skillType) } ?: throw RuntimeException()

    val nextVal = if (downgrade) {
      skillType.getPrevValue(
        actualSkill,
        if (min < 0) null else min,
        if (max < 0) null else max,
      )
    } else {
      skillType.getNextValue(actualSkill, if (min < 0) null else min, if (max < 0) null else max)
    }

    if (nextVal != null && Player.set(player, actualSkill, nextVal)) {
      maybeNotify(player, notify, nextVal.toString())
    }
  }

  companion object {
    protected const val NO_VALUE = -1.0
  }
}
